File: DesignPatterns/01_GOF23/DP_examples/Python/03_Behavioral/Command/Invoker.py
# 命令调用者


class Invoker:
    def __init__(self):
        self.commands = []
    
    def store_command(self, command):
        self.commands.append(command)
    
    def remove_command(self, command):
        if not self.commands:
            return False
        if command in self.commands:
            self.commands.remove(command)
            return True
        return False
    
    @property
    def count(self):
        return len(self.commands)
    
    def invoke(self):
        for command in self.commands:
            command.execute()

File: DesignPatterns/01_GOF23/DP_examples/Rust/03_Behavioral/Command/invoker.rs
// 命令调用者实现

// 导入命令接口和类型别名
use crate::command::CommandPtr;
use std::sync::Arc;

// 命令调用者
pub struct Invoker {
    commands: Vec<CommandPtr>,
}

impl Invoker {
    pub fn new() -> Self {
        Invoker {
            commands: Vec::new(),
        }
    }
    
    pub fn store_command(&mut self, command: &CommandPtr) {
        self.commands.push(command.clone());
    }
    
    pub fn remove_command(&mut self, command: &CommandPtr) -> bool {
        if self.commands.is_empty() {
            return false;
        }
        
        // 使用Arc的指针比较来找到并移除命令
        for i in 0..self.commands.len() {
            if Arc::ptr_eq(&self.commands[i], command) {
                self.commands.remove(i);
                return true;
            }
        }
        false
    }
    
    pub fn count(&self) -> usize {
        self.commands.len()
    }
    
    pub fn invoke(&self) {
        for command in &self.commands {
            command.execute();
        }
    }
}
